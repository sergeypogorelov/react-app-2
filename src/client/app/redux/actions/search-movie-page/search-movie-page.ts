import {
  SEARCH_MOVIES,
  SEARCH_MOVIES_PENDING,
  SEARCH_MOVIES_FULFILLED,
  SEARCH_MOVIES_REJECTED,
  CHANGE_SEARCH,
  CHANGE_SEARCH_BY,
  CHANGE_SORT_BY,
} from '../action-types';

import { GLOBALS } from '../../../core/constants/globals';

import { SearchBy } from '../../../core/enums/search-by.enum';
import { SortBy } from '../../../core/enums/sort-by.enum';
import { SortOrder } from '../../../core/enums/sort-order.enum';

import { DataRequestQueryParams } from '../../../core/interfaces/common/data-request-query-params.interface';
import { MoviesResponse } from '../../../core/interfaces/movie/movies-response.interface';

import { moviesApi } from '../../../core/api/movies/movies-api.class';

export const searchMovies = (
  search: string,
  searchBy?: SearchBy,
  sortBy?: SortBy
): SearchMoviesAction => {
  if (!search) {
    throw new Error('Search is not specified.');
  }

  if (!searchBy) {
    searchBy = SearchBy.Title;
  }

  if (!sortBy) {
    sortBy = SortBy.ReleaseData;
  }

  const params: DataRequestQueryParams = {
    search,
    searchBy,
    sortBy,
    limit: GLOBALS.moviesPerPage,
    sortOrder: SortOrder.Desc,
  };

  return {
    type: SEARCH_MOVIES,
    payload: moviesApi.getAll(params),
  };
};

export const changeSearch = (newSearch: string): ChangeSearchAction => {
  return {
    type: CHANGE_SEARCH,
    payload: newSearch,
  };
};

export const changeSearchBy = (newSearchBy: SearchBy): ChangeSearchByAction => {
  return {
    type: CHANGE_SEARCH_BY,
    payload: newSearchBy,
  };
};

export const changeSortBy = (newSortBy: SortBy): ChangeSortByAction => {
  return {
    type: CHANGE_SORT_BY,
    payload: newSortBy,
  };
};

export interface SearchMoviesAction {
  type: typeof SEARCH_MOVIES;
  payload: Promise<MoviesResponse>;
}

export interface SearchMoviesPendingAction {
  type: typeof SEARCH_MOVIES_PENDING;
}

export interface SearchMoviesFulfilledAction {
  type: typeof SEARCH_MOVIES_FULFILLED;
  payload: MoviesResponse;
}

export interface SearchMoviesRejectedAction {
  type: typeof SEARCH_MOVIES_REJECTED;
}

export interface ChangeSearchAction {
  type: typeof CHANGE_SEARCH;
  payload: string;
}

export interface ChangeSearchByAction {
  type: typeof CHANGE_SEARCH_BY;
  payload: SearchBy;
}

export interface ChangeSortByAction {
  type: typeof CHANGE_SORT_BY;
  payload: SortBy;
}

export type SearchMoviePageActionTypes =
  | SearchMoviesAction
  | SearchMoviesPendingAction
  | SearchMoviesFulfilledAction
  | SearchMoviesRejectedAction
  | ChangeSearchAction
  | ChangeSearchByAction
  | ChangeSortByAction;
