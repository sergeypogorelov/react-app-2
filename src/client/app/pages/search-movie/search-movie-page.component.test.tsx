import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';

import { SearchMoviePage } from './search-movie-page.component';
import { SearchBy } from '../../core/enums/search-by.enum';
import { SortBy } from '../../core/enums/sort-by.enum';

const mockStore = configureStore();

test('SearchMoviePage component should render correctly', () => {
  const store = mockStore({
    searchMoviePage: {
      search: '',
      searchBy: SearchBy.Title,
      sortBy: SortBy.ReleaseData,
      movies: [],
      totalMovies: 0,
    },
  });

  const json = renderer
    .create(
      <Provider store={store}>
        <SearchMoviePage />
      </Provider>
    )
    .toJSON();

  expect(json).toMatchSnapshot();
});
