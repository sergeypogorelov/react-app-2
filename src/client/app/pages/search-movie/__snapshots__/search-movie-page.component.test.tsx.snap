// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchMoviePage component should render correctly 1`] = `
<div
  className="search-movie-page"
>
  <div
    className="header"
  >
    <div
      className="header-logo"
    >
      <p
        className="title-main m-0"
      >
        <span
          className="title-main-prefix"
        >
          netflix
        </span>
        <span>
          roulette
        </span>
      </p>
    </div>
    <div
      className="header-content"
    >
      <h2
        className="header-title"
        id="searchLabel"
      >
        FIND YOUR MOVIE
      </h2>
      <form
        className="movie-search"
        onSubmit={[Function]}
      >
        <input
          aria-labelledby="searchLabel"
          className="movie-search-control"
        />
        <button
          className="movie-search-button"
        >
          Search
        </button>
      </form>
      <div
        className="switch-container d-flex align-items-center"
      >
        <label
          className="switch-container-label my-0"
        >
          SEARCH BY
        </label>
        <div
          className="switch d-flex"
        >
          <button
            className="switch-item switch-item-left active"
            onClick={[Function]}
            type="button"
          >
            Title
          </button>
          <button
            className="switch-item switch-item-right"
            onClick={[Function]}
            type="button"
          >
            Genre
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="info-container d-flex align-items-center justify-content-between"
  >
    <p
      className="info-text"
    >
      0
       movies found
    </p>
    <div
      className="switch-container d-flex align-items-center"
    >
      <label
        className="switch-container-label my-0"
      >
        SORT BY
      </label>
      <div
        className="switch d-flex"
      >
        <button
          className="switch-item switch-item-left active"
          onClick={[Function]}
          type="button"
        >
          Release Date
        </button>
        <button
          className="switch-item switch-item-right"
          onClick={[Function]}
          type="button"
        >
          Rating
        </button>
      </div>
    </div>
  </div>
  <div
    className="movies-list"
  >
    <h3
      className="movies-list-empty"
    >
      No movies found
    </h3>
  </div>
</div>
`;
